### DISCUSSION TITLE
extract LAB values of a LAB TIFF rectangle

### DISCUSSION THREAD

**Discussion by SylvainArd** [ImageMagick/ImageMagick#5059] [Help]
Hello,
I would extract LAB values of a LAB TIFF rectangle. The TIFF is LAB encoded.
Thank you
Best regards

**Comment by snibgo** [comment]
I showed a method on your other thread at https://github.com/ImageMagick/ImageMagick/discussions/5054

That gives the values in a nominal range of 0 to QuantumLevel. If you want a nominal range of 0 to 1, then divide by QuantumLevel:
```
magick out.tiff -evaluate Divide %[fx:QuantumRange] ftxt:
```
The values for the a* and b* channels have the null point at 0.5, so you might want to subtract 0.5.

**Comment by SylvainArd** [comment]
thank you and for L ?
Sylvain Ard
0549507724
0778380991
***@***.***
http://sylvain-ard.fr
Entreprise individuelle SIRET : 80079243400022
Appt 26 Bât A Résidence Le Patio
83 rue de la Bugellerie
86000 Poitiers

**Comment by snibgo** [comment]
The null point of L* is at 0.0.

**Comment by SylvainArd** [comment]
OK thank you and for a rectangle I put -crop 100x300+30+40 where 100 is the
width, 300 the height, 30 the x of top left pixel and 40 the y of top left
pixel, isn't it and the 0 of y is at top or bottom of image ?

Sylvain Ard
0549507724
0778380991
***@***.***
http://sylvain-ard.fr
Entreprise individuelle SIRET : 80079243400022
Appt 26 Bât A Résidence Le Patio
83 rue de la Bugellerie
86000 Poitiers

**Comment by snibgo** [comment]
Yes. I suggest "+repage" after the "-crop".

**Comment by SylvainArd** [comment]
yes of course ! and for my last question please ?
Sylvain Ard
0549507724
0778380991
***@***.***
http://sylvain-ard.fr
Entreprise individuelle SIRET : 80079243400022
Appt 26 Bât A Résidence Le Patio
83 rue de la Bugellerie
86000 Poitiers

**Comment by snibgo** [comment]
y=0 at the top.

**Comment by SylvainArd** [comment]
ok thank you very much and the rest is right ?
Sylvain Ard
0549507724
0778380991
***@***.***
http://sylvain-ard.fr
Entreprise individuelle SIRET : 80079243400022
Appt 26 Bât A Résidence Le Patio
83 rue de la Bugellerie
86000 Poitiers

**Comment by snibgo** [comment]
I wasn't clear when I wrote:
> The values for the a* and b* channels have the null point at 0.5, so you might want to subtract 0.5.

This is true when we convert from sRGB to CIELab with "-colorspace Lab". For example, we convert a colour close to mid-gray:
```
magick xc:#567 -colorspace Lab -evaluate Divide %[fx:QuantumRange] ftxt:
0,0:0.42407786,0.4916091,0.45441234
```
For neutral grays, a* and b* are around 0.5.

However, it is *not* true when we use profiles.
```
magick xc:#567 -profile sRGB.icc -profile Lab-D50-Identity-elle-V2.icc -evaluate Divide %[fx:QuantumRange] ftxt:
0,0:0.4170956,-0.0099724234,-0.049999982
```
For neutral grays, a* and b* are around 0.0.

**Comment by SylvainArd** [comment]
why it is not the same ?
Sylvain Ard
0549507724
0778380991
***@***.***
http://sylvain-ard.fr
Entreprise individuelle SIRET : 80079243400022
Appt 26 Bât A Résidence Le Patio
83 rue de la Bugellerie
86000 Poitiers

**Comment by snibgo** [comment]
"-colorspace " was written decades ago, when IM stored pixel values as positive integers from zero to QuantumLevel. So a* and b* values needed to have an offset added so they were not negative.

"-profile" is more recent. Lab profiles need HDRI (floating-point) versions of IM, to cope with the negative values.

**Comment by SylvainArd** [comment]
OK but normally L vary from 0 to 100 and A and B vary from -127 to 127,
what is the match please ?
Sylvain Ard
0549507724
0778380991
***@***.***
http://sylvain-ard.fr
Entreprise individuelle SIRET : 80079243400022
Appt 26 Bât A Résidence Le Patio
83 rue de la Bugellerie
86000 Poitiers

**Comment by snibgo** [comment]
Multiply just those channels by the appropriate factors. For example (Windows BAT syntax):
```
magick ^
  in.tiff ^
  -channel 0 -evaluate Multiply %%[fx:100/QuantumRange] ^
  -channel 1,2 -evaluate Multiply %%[fx:255/QuantumRange] ^
 +channel ^
  ftxt:
```

**Comment by SylvainArd** [comment]
Ok thank you but for a and b it would be 
254/quantumrange-127
Envoyé de mon iPhone
